urlcdd0c1ab49_emiln_JHouseFireExplorer_tgz-pJ8-dk_brics_automaton_DatatypesJ8/src/dk/brics/automaton/Datatypes.java

```diff
@@ -0,0 +1,880 @@
+	 * <tr><td><tt>Sc</tt></td><td><a target="_top" href="http://www.unicode.org/Public/3.1-Update/UnicodeData-3.1.0.txt">Sc</a> category from Unicode 3.1</td></tr>
+	 * <tr><td><tt>Sk</tt></td><td><a target="_top" href="http://www.unicode.org/Public/3.1-Update/UnicodeData-3.1.0.txt">Sk</a> category from Unicode 3.1</td></tr>
+	 * <tr><td><tt>So</tt></td><td><a target="_top" href="http://www.unicode.org/Public/3.1-Update/UnicodeData-3.1.0.txt">So</a> category from Unicode 3.1</td></tr>
+	 * <tr><td><tt>S</tt></td><td><a target="_top" href="http://www.unicode.org/Public/3.1-Update/UnicodeData-3.1.0.txt">S</a> category from Unicode 3.1</td></tr>
+	 * <tr><td><tt>Cc</tt></td><td><a target="_top" href="http://www.unicode.org/Public/3.1-Update/UnicodeData-3.1.0.txt">Cc</a> category from Unicode 3.1</td></tr>
+	 * <tr><td><tt>Cf</tt></td><td><a target="_top" href="http://www.unicode.org/Public/3.1-Update/UnicodeData-3.1.0.txt">Cf</a> category from Unicode 3.1</td></tr>
+	 * <tr><td><tt>Co</tt></td><td><a target="_top" href="http://www.unicode.org/Public/3.1-Update/UnicodeData-3.1.0.txt">Co</a> category from Unicode 3.1</td></tr>
+	 * <tr><td><tt>Cn</tt></td><td><a target="_top" href="http://www.unicode.org/Public/3.1-Update/UnicodeData-3.1.0.txt">Cn</a> category from Unicode 3.1</td></tr>
+	 * <tr><td><tt>C</tt></td><td><a target="_top" href="http://www.unicode.org/Public/3.1-Update/UnicodeData-3.1.0.txt">C</a> category from Unicode 3.1</td></tr>
+	 * </table>
+	 * <p>Loaded automata are cached in memory.
+	 * @param name name of automaton
+	 * @return automaton
+	 */
+	public static Automaton get(String name) {
+		Automaton a = automata.get(name);
+		if (a == null) {
+			a = load(name);
+			automata.put(name, a);
+		}
+		return a;
+	}
+	
+	/**
+	 * Checks whether the given string is the name of a Unicode block (see {@link #get(String)}).
+	 */
+	public static boolean isUnicodeBlockName(String name) {
+		return unicodeblock_names.contains(name);
+	}
+	
+	/**
+	 * Checks whether the given string is the name of a Unicode category (see {@link #get(String)}).
+	 */
+	public static boolean isUnicodeCategoryName(String name) {
+		return unicodecategory_names.contains(name);
+	}
+	
+	/**
+	 * Checks whether the given string is the name of an XML / XML Schema automaton (see {@link #get(String)}).
+	 */
+	public static boolean isXMLName(String name) {
+		return xml_names.contains(name);
+	}
+	
+	/**
+	 * Checks whether a given automaton is available.
+	 * @param name automaton name
+	 * @return true if the automaton is available
+	 */
+	public static boolean exists(String name) {
+		try {
+			Datatypes.class.getClassLoader().getResource(name + ".aut").openStream().close();
+		} catch (IOException e) {
+			return false;
+		}
+		return true;
+	}
+
+	private static Automaton load(String name) {
+		try {
+			URL url = Datatypes.class.getClassLoader().getResource(name + ".aut");
+			return Automaton.load(url.openStream());
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the resource returned by the following method call: *openStream*. Currently, there are execution paths that do not contain closure statements, for example, when Automaton.load throws an exception. Ensure the object returned by *openStream()* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).